File: leetcode_python/101. Symmetric Tree.py
class Solution:
    def isSymmetric(self, root: TreeNode) -> bool:
        def match(l, r):
            if not l and not r: return True
            if not l or not r: return False
            return l.val == r.val and \
                match(l.left, r.right) and \
                match(l.right, r.left)
        return match(root, root)

File: leetcode_java/101. Symmetric Tree.java
class Solution {
    public boolean isSymmetric(TreeNode root) {
        return match(root, root);
    }
    private boolean match(TreeNode l, TreeNode r) {
        if (l == null && r == null) return true;
        if (l == null || r == null) return false;
        return l.val == r.val && 
            match(l.left, r.right) && 
            match(l.right, r.left);
    }
}